texture.h

```diff
@@ -0,0 +1,63 @@
+/* Copyright (C) 2013-2014 Michal Brzozowski (rusolis@poczta.fm)
+
+   This file is part of KeeperRL.
+
+   KeeperRL is free software; you can redistribute it and/or modify it under the terms of the
+   GNU General Public License as published by the Free Software Foundation; either version 2
+   of the License, or (at your option) any later version.
+
+   KeeperRL is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+   even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License along with this program.
+   If not, see http://www.gnu.org/licenses/ . */
+
+#pragma once
+
+#include "util.h"
+#include "sdl.h"
+#include "color.h"
+#include "file_path.h"
+
+class Texture {
+  public:
+  Texture();
+  Texture(const Texture&) = delete;
+  Texture(Texture&&);
+  Texture& operator=(Texture&&);
+  Texture(const FilePath& path);
+  Texture(const FilePath& path, int px, int py, int kx, int ky);
+  Texture(Color, int width, int height);
+  explicit Texture(SDL::SDL_Surface*);
+  ~Texture();
+
+  static optional<Texture> loadMaybe(const FilePath&);
+  optional<SDL::GLenum> loadFromMaybe(SDL::SDL_Surface*);
+
+  Vec2 getSize() const {
+    return size;
+  }
+  Vec2 getRealSize() const {
+    return realSize;
+  }
+  unsigned getTexId() const {
+    return texId;
+  }
+  const optional<FilePath>& getPath() const {
+    return path;
+  }
+
+  static SDL::SDL_Surface* createSurface(int w, int h);
+  static SDL::SDL_Surface* createPowerOfTwoSurface(SDL::SDL_Surface*);
+
+  private:
+  void addTexCoord(int x, int y) const;
+
+  // Note: it's better to use 0 instead of optional:
+  // OpenGL object id's are always > 0
+  SDL::GLuint texId = 0;
```

Comment:
I like optional because it fails early if you try to use it when uninitialized. What's the advantage of using 0 as special value?

Reply:
I see following cases:
1. Texture is properly initialized with some data (from file, SDL surface or in some other way); texId is properly initialized
2. Texture is constructed with default constructor; texId == 0
3. Texture is constructed and then initialized by loadFromMaybe; after that it may be left in valid or invalid state; If it's invalid then Texture shouldn't be used at all (but texId is still initialized)
4. There was some problem when constructing texture, we fail on a check and don't return; Texture is never constructed

Optional can help us only in case 2. But, we have to ask: is it legal to use a texture initialized with a default constructor? In OpenGL ID '0' is treated specially in case of textures: it's used for default texture which is simply black. It's perfectly legal to bind it and use it. So IMO it's OK to allow usage of such texture. In other case it would be better not to provide a default constructor at all or make it private, only to be used in functions like loadFromMaybe.

Reply:
There is no use case in the game for using an uninitialized (black) texture, so I prefer the game to crash if you try to use one by accident. The alternative is having "why the hell is this sprite not being drawn" kind of bugs. Similar case with using a moved-from texture.

Ideally there shouldn't be a default constructor, but the way textures are stored in EnumMaps and loaded from files, you would have to store optional<Texture> everywhere. Which isn't bad, but serves the exact same logic as having a default "none" texture like now, if you think about it.

Reply:
OK, makes sense.